### ISSUE TITLE
Check if property is used before assignment

### ISSUE THREAD

**Issue by koral--** [detekt/detekt#749]
For example in this case:
```    
class Foo {
    val number
        get() = if (isValid) 1 else 0
    val list = listOf(number)
    val isValid = true
}
println(Foo().list)
```
`0` will be printed because `isValid` is `false` when getter is invoked first time. If you move `val isValid = true` at least one line above, `1` will be printed.

**Comment by bubenheimer** [comment]
I'd think adding this rule would make a big difference to anyone using Kotlin, as this problem is easy to introduce and miss, without any current automated way of detecting it AFAIK, not even runtime checks. Or if there is a tool out there to find this problem, please correct my ignorance!

The general case may be non-trivial to solve without the potential for false positives. Still, I'd imagine a rule that is disabled by default, which just validates textual "definition before use" may be possible to add easily. I'd think that "definition before use" should be a strongly encouraged practice in Kotlin because of this issue, whenever possible.

There are common special cases that should be easier to detect reliably, in particular use of not-yet-defined properties in stdlib scope functions (let, run, also, apply, with). That's the one that got me.

**Comment by arturbosch** [comment]
@bubenheimer can you please add your example as a comment.

**Comment by bubenheimer** [comment]
Example with stdlib scope function:

    class Bar {
        val map: GoogleMap = GoogleMap()

        init {
            map.apply {
                // non-null clickListener is actually null,
                // so map click listener is set to null
                // without error or warning at compile time or runtime
                setOnMapClickListener(clickListener)

                // Set a dozen more listeners below
                // ...
            }
        }

        val clickListener: OnMapClickListener = OnMapClickListener {
            print("Map clicked")
        }

        // Define a dozen more listeners here
        // ...
    }

**Comment by BraisGabin** [comment]
I like the idea of "textually later". And I see this as an error prone rule. It could not be an error but it can create it. And you usually can arrange the code to fix it. If it's not possible and you know that it's not a problem you can always suppress it.
